frame/support/src/storage/types/counted_nmap.rs

```diff
@@ -0,0 +1,1445 @@
+		<Self as MapWrapper>::Map::module_prefix()
+	}
+	fn storage_prefix() -> &'static [u8] {
+		<Self as MapWrapper>::Map::storage_prefix()
+	}
+}
+
+impl<Prefix, Key, Value, QueryKind, OnEmpty, MaxValues>
+	CountedStorageNMap<Prefix, Key, Value, QueryKind, OnEmpty, MaxValues>
+where
+	Prefix: CountedStorageNMapInstance,
+	Key: super::key::KeyGenerator,
+	Value: FullCodec,
+	QueryKind: QueryKindTrait<Value, OnEmpty>,
+	OnEmpty: Get<QueryKind::Query> + 'static,
+	MaxValues: Get<Option<u32>>,
+{
+	/// Get the storage key used to fetch a value corresponding to a specific key.
+	pub fn hashed_key_for<KArg: EncodeLikeTuple<Key::KArg> + TupleToEncodedIter>(
+		key: KArg,
+	) -> Vec<u8> {
+		<Self as MapWrapper>::Map::hashed_key_for(key)
+	}
+
+	/// Does the value (explicitly) exist in storage?
+	pub fn contains_key<KArg: EncodeLikeTuple<Key::KArg> + TupleToEncodedIter>(key: KArg) -> bool {
+		<Self as MapWrapper>::Map::contains_key(key)
+	}
+
+	/// Load the value associated with the given key from the map.
+	pub fn get<KArg: EncodeLikeTuple<Key::KArg> + TupleToEncodedIter>(
+		key: KArg,
+	) -> QueryKind::Query {
+		<Self as MapWrapper>::Map::get(key)
+	}
+
+	/// Try to get the value for the given key from the map.
+	///
+	/// Returns `Ok` if it exists, `Err` if not.
+	pub fn try_get<KArg: EncodeLikeTuple<Key::KArg> + TupleToEncodedIter>(
+		key: KArg,
+	) -> Result<Value, ()> {
+		<Self as MapWrapper>::Map::try_get(key)
+	}
+
+	/// Store or remove the value to be associated with `key` so that `get` returns the `query`.
+	pub fn set<KArg: EncodeLikeTuple<Key::KArg> + TupleToEncodedIter>(
+		key: KArg,
+		query: QueryKind::Query,
+	) {
+		let option = QueryKind::from_query_to_optional_value(query);
+		if option.is_none() {
+			CounterFor::<Prefix>::mutate(|value| value.saturating_dec());
+		}
+		<Self as MapWrapper>::Map::set(key, QueryKind::from_optional_value_to_query(option))
+	}
+
+	/// Take a value from storage, removing it afterwards.
+	pub fn take<KArg: EncodeLikeTuple<Key::KArg> + TupleToEncodedIter>(
+		key: KArg,
+	) -> QueryKind::Query {
+		let removed_value =
+			<Self as MapWrapper>::Map::mutate_exists(key, |value| core::mem::replace(value, None));
+		if removed_value.is_some() {
+			CounterFor::<Prefix>::mutate(|value| value.saturating_dec());
+		}
+		QueryKind::from_optional_value_to_query(removed_value)
+	}
+
+	/// Swap the values of two key-pairs.
+	pub fn swap<KOther, KArg1, KArg2>(key1: KArg1, key2: KArg2)
+	where
+		KOther: KeyGenerator,
+		KArg1: EncodeLikeTuple<Key::KArg> + TupleToEncodedIter,
+		KArg2: EncodeLikeTuple<KOther::KArg> + TupleToEncodedIter,
+	{
+		<Self as MapWrapper>::Map::swap::<KOther, _, _>(key1, key2)
+	}
+
+	/// Store a value to be associated with the given keys from the map.
+	pub fn insert<KArg, VArg>(key: KArg, val: VArg)
+	where
+		KArg: EncodeLikeTuple<Key::KArg> + EncodeLike<Key::KArg> + TupleToEncodedIter,
+		VArg: EncodeLike<Value>,
+	{
+		if !<Self as MapWrapper>::Map::contains_key(Ref::from(&key)) {
```

Comment:
doesn't insert return any kind of flag which indicates if value was present before?

Reply:
I have followed an original `NMap` interface, and we don't have any return value there. I don't know actually which approach should be better.  If you are insist that we should return a flag, don't see any problem to do it, I will update it. So what do think, should I fix it ?

https://github.com/Mr-Leshiy/substrate/blob/2a5c7781c16b845e59c4d269b51f3f99228676e6/frame/support/src/storage/types/nmap.rs#L171

Reply:
That's unfortunate. If `NMap::insert` had a flag indicating if key was present before (by analogy with https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.insert), then we wouldn't need to call `contains_key`.

I think we should consider changing the API, but this can be a separate issue/PR.

Reply:
I agree with you, it would be better to update an interface. But I would rather describe it in the issue first and update it in the separate PR, because it will needed to modify and revise other parts of the Storages implementation. Would you create it, or I should do it ?

Reply:
I'll open an issue.